iMAD Tutorial Part 1

mortcanty opened the pull request:
Since it is a bit long, I want to split the tutorial into three parts. This is the first. I suppose this PR doesn't close Issue #571 since there are two more to come. Thanks, Mort

mortcanty commented:
Hi and Happy New Year. Know you're all busy, but any timeline for a review of this PR? Thanks.

gorelick-google commented:
Hi Mort,

I think you requested me to review it, but I just haven't had the time (and
colab isn't my thing).
You might want to try someone else?

On Tue, Jan 3, 2023 at 10:56 AM Dr. Mort Canty ***@***.***>
wrote:

> Hi and Happy New Year. Know you're all busy, but any timeline for a review
> of this PR? Thanks.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/google/earthengine-community/pull/583#issuecomment-1369575329>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AM25RFT35KXYYRVOABLI5MTWQPZURANCNFSM6AAAAAASLFO3YI>
> .
> You are receiving this because you were assigned.Message ID:
> ***@***.***>
>

mortcanty commented:
Thanks, Noel. As I understand the procedure so far, Gino assigned you to the review. Since you're not available, Justin Braaten would be a good choice as he reviewed the four SAR Colab tutorials. Gino, can you ask him or should I?

mortcanty commented:
This seems to be taking a while :-) 
At Justin's suggestion I'm using Tyler Erickson's experimental ee_jupyter package extensively in  the Tutorial, so maybe I could suggest either Justin or Tyler as alternative to Noel as reviewer for this PR.

jdbcode commented:
Hi Mort, I'll do the review!

mortcanty commented:
Great, thanks Justin.

mortcanty commented:
Justin, my partner Allan Nielsen and I have been working for a while on Part 2 of the proposed Tutorial (the main part). Your opinion/comments on Part 1 would be a big help to get it right. Any timeline for your review? Thanks.

jdbcode commented:
Hi Mort, I've blocked out time this afternoon to look.

jdbcode commented:
I'm try'n, Mort! Internal priorities are blocking. I'm planning on doing a less intensive review. I'll mostly be running the notebook to look for errors and a light grammar edit for anything glaring. You're the expert on this, so I'll most be helping to navigate getting it from GitHub to Google. 

One potential hitch is that our live outputs will not currently work with the ee_jupyter library. I don't think that should stop us from moving this forward (plus, I really like the library). I need to check if it will break anything though, or simply not show the outputs in the DevSite page.

mortcanty commented:
Thanks a lot, Justin. I rather had the impression that you (and your colleagues at GEE) were very busy. Your help is much appreciated. I like ee_jupyter as well. As I recall, there was a similar problem with Part 4 of my Sentinel-1 Tutorial, in which I used IPyLeaflet widgets and maps. But you found a workaround/compromise.

jdbcode commented:
Okay, I was able to dig into this today. As you suggested, we should follow the path we took on [part 4 of the SAR change detection series](https://github.com/google/earthengine-community/tree/master/tutorials/detecting-changes-in-sentinel-1-imagery-pt-4): convert the notebook to markdown (using `jupyter nbconvert --to markdown`) for display on Earth Engine DevSite page and have it link to the Colab notebook that you host in your own GitHub repo (in the future it would be great if the markdown and notebook files lived together in the same github.com/google/earthengine-community folder, but let's not let the current handling of .md and .ipynb in same folder stop us from publishing).

So... I think these are the steps:

- [ ] I'll do minor review on the notebook file in this pull request
- [ ] You make any suggested changes from the review, commit them here
- [ ] I'll check changes and say "looks good"
- [ ] You move/copy the "looks good" notebook file into a repo in your GitHub account (similar to [SAR part 4](https://github.com/mortcanty/eesarseq/blob/main/src/Tutorial_pt_4.ipynb))
- [ ] I'll send a markdown version of it to you and you commit it to this pull request and remove the notebook
- [ ] I'll approve the pull request and make sure that it publishes correctly

Now that I've had a serious look and sized up the situation, I feel confident I can complete the review on Monday morning.

mortcanty commented:
Right, can do. It is a bit of a bummer that the publication has to be so roundabout. There are two more parts to come, although part 3 will take a while and I will only request it be pulled if we think it will contribute something of value to normalization/harmonization. Thanks and sorry this bites so much into you time.

jdbcode commented on tutorials/imad-tutorial-pt1/index.ipynb:
```diff
@@ -0,0 +1,2078 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #2.    aoi = ee.Geometry(aois.geometries().get(26))
</pre><p> </p><p>This is sort of a bad convention in GEE, is their a property you can filter on e.g., <code>aois.filter(...).geometry()</code>, instead of using <code>aois.geometries.get(index)</code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/google/earthengine-community/pull/583/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

mortcanty replied:
Listing the properties of aois all I get is ['system:version', 'system:id', 'system:asset_size']. Left as is

jdbcode replied:
Acknowledged

jdbcode commented on tutorials/imad-tutorial-pt1/index.ipynb:
```diff
@@ -0,0 +1,2078 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #6.                                   .filter(ee.Filter.contains(rightValue=aoi,leftField='.geo')) 
</pre><p> </p><p>Are both region filters necessary (i.e. <code>filterBounds</code> and <code>ee.Filter.contains</code>)? My guess is you can just use <code>ee.Filter.contains</code>. This applies to <code>im1</code> and <code>im2</code>  </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/google/earthengine-community/pull/583/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='13'/>

mortcanty replied:
Yes, but it takes an orders of magnitude longer to find regions and then times out when trying to plot the them. Left unchanged.

jdbcode replied:
Interesting. Acknowledged no change.

jdbcode commented on tutorials/imad-tutorial-pt1/index.ipynb:
```diff
@@ -0,0 +1,2078 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #17.            timestamp = im1.get('system:time_start').getInfo()
</pre><p> </p><p>You could shorted these time formatting steps using this:</p><p> </p><p><code>im1.date().format('E MMM dd HH:mm:ss YYYY')</code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/google/earthengine-community/pull/583/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='13'/>

jdbcode commented:
Mort, I've completed some minor edits:

- Changed filename to .ipynb (so that ReviewNB tool would work)
- Fixed obvious whitespace and grammar issues
- Light review of code (see ReviewNB comments above)

Next step is for you to:

- [ ] make any suggested changes from the review, commit them here

Sorry the process is a made a little janky by non-supported Python libraries. I'll bring this up as something to fix.

jdbcode commented:
Mort, I've emailed you the final notebook (from [fa4c418](https://github.com/google/earthengine-community/pull/583/commits/fa4c41867095d57f9f1e433041cadfa5d60a3e64)) and the Markdown conversion along with the next steps. After I see the markdown version committed here, I'll verify it and then it should be ready to merge.

mortcanty commented:
Great. A bit past my bedtime here, so I'll do it tomorrow morning. Cheers.

mortcanty commented:
OK, I created my new repo https://github.com/mortcanty/eeimad with the ipynb tutorial notebook, removed index.ipynb from the pull request and replaced with index.md, with link to the tutorial notebook. After commit, seems to be having trouble with the two authors format.

jdbcode reviewed (approved):
All clear for merge!
It'll get pulled to Google, I'll check the formatting there, apply remaining format fixes (if any needed), and let you know when it's live (not more than a week).

mortcanty commented:
Fantastisch (as they say over here). Once again many thanks for your help and engagement!

jdbcode commented:
It's live!
https://developers.google.com/earth-engine/tutorials/community/imad-tutorial-pt1

Thanks for supporting the EE developer community by sharing your expertise!

Note that in the page linked above I had to remove the second author because of improper formatting/hyperlinking. We need to fix our parser, but I didn't want to hold this up any longer. Once the parser is fixed, I'll add the second author back on.
